### PR TITLE
Updates gait analysis

### PR DESCRIPTION
**Pull request by antoinefalisse**
@suhlrich This PR does a couple of things related to gait analysis:

1. It adds support for trimming trials (start and/or end). For the gait module, I suggest we trim 0.5 at the end.
2. It detects the walking direction (+/- x) and adjusts the _Subtract sacrum from foot_ accordingly. It was necessary to have the analysis work when the subject is walking towards negative x (cameras in the back). What I did is very basic, I looked at the pelvis markers (asis vs psis) and assumed that was the walking direction.
3. It adds a +1 (we discussed that) such that the normalized cycles include the second HS.
4. It throws an error instead of a warning if the gait event order is wrong. I rather have no results than bad results.

Remark:
- The algorithm does not work well when the subject is turning (like you would for 6MWT). I am thinking of adding a **?** on the frontend that describes what to do when using the gait module.

### CODE REVIEW

File: ActivityAnalyses/gait_analysis.py [hunk 1]
```diff
@@ -732,6 +761,17 @@ def detect_correct_order(rHS, rTO, lHS, lTO):
         l_toe_rel_x = (
             self.markerDict['markers']['L_toe_study'] - 
             self.markerDict['markers']['L.PSIS_study'])[:,0]
+        
+        # Identify which direction the subject is walking.
```
Comment (suhlrich) [inline ActivityAnalyses/gait_analysis.py:765]: Good catch. I suggest we go even more general:
1) find heading vector from mid-psis to mid-asis; project onto floor (ie make y=0)
2) dot projected heading vector with the l_calc_rel vecor (get rid of the [:,0] before dot.
3) the result should function how the r_toe_rel_x, etc... is already functioning.

Comment (suhlrich) [reply ActivityAnalyses/gait_analysis.py:765]: This will help if they walk off axis from the checkerboard at all

**Review by suhlrich — CHANGES_REQUESTED** [review]
Great! One comment.

File: ActivityAnalyses/gait_analysis.py (hunk 1 above)
Comment (antoinefalisse) [reply ActivityAnalyses/gait_analysis.py:765]: @suhlrich I am not super clear on what you wrote. Is this what you mean?

```
      r_calc_rel = (
          self.markerDict['markers']['r_calc_study'] - 
          self.markerDict['markers']['r.PSIS_study'])
      r_calc_rel_x = r_calc_rel[:,0]
      r_toe_rel = (
          self.markerDict['markers']['r_toe_study'] - 
          self.markerDict['markers']['r.PSIS_study'])
      r_toe_rel_x = r_toe_rel[:,0]

      # Repeat for left.
      l_calc_rel = (
          self.markerDict['markers']['L_calc_study'] - 
          self.markerDict['markers']['L.PSIS_study'])
      l_calc_rel_x = l_calc_rel[:,0]
      l_toe_rel = (
          self.markerDict['markers']['L_toe_study'] - 
          self.markerDict['markers']['L.PSIS_study'])
      l_toe_rel_x = l_toe_rel[:,0]
      
      # Identify which direction the subject is walking.
      mid_psis = (self.markerDict['markers']['r.PSIS_study'] + self.markerDict['markers']['L.PSIS_study'])/2
      mid_asis = (self.markerDict['markers']['r.ASIS_study'] + self.markerDict['markers']['L.ASIS_study'])/2
      mid_dir = mid_asis - mid_psis
      mid_dir_floor = np.copy(mid_dir)
      mid_dir_floor[:,1] = 0
      mid_dir_floor = mid_dir_floor / np.linalg.norm(mid_dir_floor,axis=1,keepdims=True)
      l_calc_rel_floor = np.copy(l_calc_rel)
      l_calc_rel_floor[:,1] = 0
      l_calc_rel_floor = l_calc_rel_floor / np.linalg.norm(l_calc_rel_floor,axis=1,keepdims=True)        
      dot_product = np.einsum('ij,ij->i', mid_dir_floor,l_calc_rel_floor)
```

How do you then use the dot_product?

Comment (suhlrich) [reply ActivityAnalyses/gait_analysis.py:765]: `      r_calc_rel = (
          self.markerDict['markers']['r_calc_study'] - 
          self.markerDict['markers']['r.PSIS_study'])

      r_toe_rel = (
          self.markerDict['markers']['r_toe_study'] - 
          self.markerDict['markers']['r.PSIS_study'])


      # Repeat for left.
      l_calc_rel = (
          self.markerDict['markers']['L_calc_study'] - 
          self.markerDict['markers']['L.PSIS_study'])

      l_toe_rel = (
          self.markerDict['markers']['L_toe_study'] - 
          self.markerDict['markers']['L.PSIS_study'])

      
      # Identify which direction the subject is walking.
      mid_psis = (self.markerDict['markers']['r.PSIS_study'] + self.markerDict['markers']['L.PSIS_study'])/2
      mid_asis = (self.markerDict['markers']['r.ASIS_study'] + self.markerDict['markers']['L.ASIS_study'])/2
      mid_dir = mid_asis - mid_psis
      mid_dir_floor = np.copy(mid_dir)
      mid_dir_floor[:,1] = 0
      mid_dir_floor = mid_dir_floor / np.linalg.norm(mid_dir_floor,axis=1,keepdims=True)
      l_calc_rel_floor = np.copy(l_calc_rel)
      l_calc_rel_floor[:,1] = 0
      l_calc_rel_floor = l_calc_rel_floor / np.linalg.norm(l_calc_rel_floor,axis=1,keepdims=True)        
      l_calc_rel_x = np.einsum('ij,ij->i', mid_dir_floor,l_calc_rel_floor)
      
      # TODO RIGHT SIDE
      
      `

Comment (antoinefalisse) [reply ActivityAnalyses/gait_analysis.py:766]: Ah I see, I adjusted. Checked the results on a few cases and very very small differences (like one frame difference causing a 1% change in symmetry). I also visualized old and new l_calc_rel_x  etc and they are essentially on top of each other, so I think we are good. Another look at the PR would be great though. Thanks.

**Review by suhlrich — APPROVED** [review]
looks great. thanks!

**Pull request merged**